### PR TITLE
MAINT: Fix bloated mismatch error percentage in array comparisons. 

### PR DESCRIPTION
**Pull request by maxwell-aladago**
Fixes [#13834]

Also related to [#14323], part of which was closed via [#14326]

### CODE REVIEW

File: numpy/testing/_private/utils.py [hunk 1]
```diff
@@ -807,7 +808,8 @@ def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
 
                     # note: this definition of relative error matches that one
                     # used by assert_allclose (found in np.isclose)
-                    max_rel_error = (error / abs(y)).max()
+                    # max_rel_error is 'inf' when y = 0
+                    max_rel_error = max(error / abs(y))
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: I think `np.max` would be better here and above

File: numpy/testing/_private/utils.py [hunk 2]
```diff
@@ -788,16 +788,17 @@ def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
         # np.ma.masked, which is falsy).
         if cond != True:
             n_mismatch = reduced.size - reduced.sum(dtype=intp)
-            percent_mismatch = 100 * n_mismatch / ox.size
+            n_elements = reduced.size + flagged.sum(dtype=intp)
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: Possibly clearer as the following, which also avoids the sum:
```python
# if we computed flags, then they are the full size of the comparison
n_elements = flagged.size if flagged.ndim != 0 else reduced.size
```

Lets see if anyone else agrees before changing.

Comment (maxwell-aladago) [reply numpy/testing/_private/utils.py]: Yes, that's better. Thanks @eric-wieser

File: numpy/testing/_private/utils.py [hunk 3]
```diff
@@ -20,7 +20,7 @@
 import pprint
 
 from numpy.core import(
-     intp, float32, empty, arange, array_repr, ndarray, isnat, array)
+     intp, float32, empty, arange, array_repr, ndarray, isnat, array, max)
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: This is risky - there are places in this code that are using the builtin `max`, and you just broken them - they are not drop in replacements for each other. I'd prefer to just spell `np.max` explicitly in the handful of places where we want that one.

File: numpy/testing/_private/utils.py [hunk 4]
```diff
@@ -19,6 +19,7 @@
 from warnings import WarningMessage
 import pprint
 
+import numpy as np
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: In other places in this file, the `import numpy as np` has been moved into function scope. I suggest we use the same approach for your changes.

File: numpy/testing/_private/utils.py [hunk 5]
```diff
@@ -788,32 +789,34 @@ def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
...
+                    n_mismatch, n_elements, percent_mismatch)]
 
             with errstate(invalid='ignore', divide='ignore'):
                 # ignore errors for non-numeric types
                 with contextlib.suppress(TypeError):
                     error = abs(x - y)
-                    max_abs_error = error.max()
-                    if error.dtype == 'object':
+                    max_abs_error = np.max(error)
+                    if hasattr(error, 'dtype') and error.dtype != 'object':
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: Nit: Perhaps `getattr(error, 'dtype', np.object_) == np.object_` which avoids repeated lookups or string coercion - and also minimizes the diff compared to the original

**Comment by eric-wieser** [comment]
Sorry for the delayed re-review. Looking pretty close now - good catch on the extra issues that appear once you fix the `max`.

**Comment by maxwell-aladago** [comment]
> Sorry for the delayed re-review. Looking pretty close now - good catch on the extra issues that appear once you fix the `max`.

Thanks @eric-wieser . The changes have been made.

**Comment by maxwell-aladago** [comment]
@charris, the `Shippable` check on this PR is failing due to `Merge Conflicts`. Can you please help me check what it is? Thank you

**Comment by seberg** [comment]
@maxwell-aladago the branch has merge conflicts and we cannot merge in any case (should be the same issue). Can you pull master, and then `git rebase master` on top of this? You will have some conflict you have to resolve probably, I hope it is not too difficult. I think it is due to a fix for arrays containing zeros that was merged very recently.

**Comment by eric-wieser** [comment]
Reopening to pick up the incoming force push

**Review by eric-wieser — COMMENTED** [review]
Would be good if we could add a test for this somehow. Patch otherwise looks good.

**Comment by maxwell-aladago** [comment]
@eric-wieser, the Shippable has been pending ever since. Can you please check it out for me?

File: numpy/testing/tests/test_utils.py [hunk 6]
```diff
@@ -564,6 +564,16 @@ def test_error_message(self):
...
+        # check the error message when x is scalar
+        x = 2
+        y = np.ones(20)
+        with pytest.raises(AssertionError) as exc_info:
+            self._assert_func(x, y)
+        msgs = str(exc_info.value).split('\n')
+        assert_equal(msgs[3], 'Mismatched elements: 20 / 20 (100%)')
+        assert_equal(msgs[4], 'Max absolute difference: 1.')
+        assert_equal(msgs[5], 'Max relative difference: 1.')
+
```
Comment (mattip) [inline numpy/testing/tests/test_utils.py]: Could you split this into its own test? All you would need I think is to add `def test_error_message2(self):` before the lines you added.

**Comment by mattip** [comment]
Thanks @maxwell-aladago

**Pull request merged**